titan-compiler/parser.lua

```diff
@@ -21,170 +47,266 @@ function defs.boolopt(x)
     return x ~= ''
 end
 
+function defs.name_exp(name)
+    return ast.Exp.Var(ast.Var.Name(name))
+end
+
 function defs.ifstat(exp, block, thens, elseopt)
-    table.insert(ast.Then.Then(exp, block), thens, 1) 
-    return ast.State.If(thens, elseopt)
+    table.insert(thens, 1, ast.Then.Then(exp, block)) 
+    return ast.Stat.If(thens, elseopt)
 end
 
 function defs.defstat(decl, exp)
-    local decl = ast.Stat.Decl(decl)
+    local declstat = ast.Stat.Decl(decl)
     local assign = ast.Stat.Assign(ast.Var.Name(decl.name), exp)
-    return decl, assign
+    return declstat, assign
 end
 
-function defs.totrue()
-    return true
+function defs.fold_binop_left(matches)
+    local lhs = matches[1]
+    for i = 2, #matches, 2 do
+        local op  = matches[i]
+        local rhs = matches[i+1]
+        lhs = ast.Exp.Binop(lhs, op, rhs)
+    end
+    return lhs
 end
 
-function defs.tofalse()
-    return false
+function defs.binop_right(lhs, op, rhs)
+    if op then
+        return ast.Exp.Binop(lhs, op, rhs)
+    else
+        return lhs
+    end
 end
 
-function defs.prefixname(name)
-    return ast.Exp.Var(ast.Var.Name(name))
+function defs.fold_unops(unops, exp)
+    for i = #unops, 1, -1 do
+        local op = unops[i]
+        exp = ast.Exp.Unop(op, exp)
+    end
+    return exp
 end
 
-function defs.varindex(prefix, suffixes, index)
-    local exp = prefix
-    for _, suffix in ipairs(suffixes) do
-        if suffix._typename == 'Call' then
-            exp = ast.Exp.Call(exp, suffix)
-        elseif suffix._typename == 'Exp' then
-            local var = ast.Var.Index(exp, suffix)
-            exp = ast.Exp.Var(var)
-        else
-            error('impossible')
-        end
+function defs.fold_suffixes(exp, suffixes)
+    for i = 1, #suffixes do
+        local suf = suffixes[i]
+        exp = suf(exp)
```

Comment:
I think we need at least a commentary explaining that suffixes captures return functions.

Reply:
We may also rename `suffix` to `applysuffix`

Reply:
I would use more meaningful names to operations. For instance, instead of 
```
e1              <- {| e2  (op1  e2 )* |}
```
I would use something like
```
e1              <- {| e2  (opOR  e2 )* |}
```
but this is just a suggestion, not a requirement for accepting the PR. :)